meta-mbl-distro/conf/distro/mbl.conf

```diff
@@ -13,7 +13,6 @@
 # - Specifying virtual providers for subsystems.
 ###############################################################################
 
-require conf/distro/mbl-partitions.conf
```

Comment:
No, conf files can't `inherit`. I'm not even sure if I'd prefer an approach like that anyway - The approach where recipes have to `inherit mbl-partitions` reduces the scope of the variables so it's easier to keep track of dependencies on partition config.

Reply:
Think in conf files you can use `INHERIT = "mbl-partitions"`.

I don't see a big issue of having a slightly bigger scope of variables.

If someone wants to use these variables within their recipe and forget to inherit this class they will have empty values which can cause build/running problems that can be difficult to identify.

For some using the mbl distro should be implicit that they will have access to those variables without needing to inherit the class to use them.

Reply:
> Think in conf files you can use `INHERIT = "mbl-partitions"`.

Ah, yes, after re-reading the `INHERIT` documentation, yes, I realize I misunderstood what it actually did.

TBH, I'm a little bit ambivalent about this. 

I don't really want users to have access to all of the variables in the class, I want them to use the interface I've provided (being able to override certain variables from conf files) and never inherit from from the class outside of meta-mbl. Otherwise assumptions that the class makes can be easily broken. If users actually did start to use these variables in ways that aren't intended then the user code could become dependent on implementation details that we might want to change later. 

On the other hand, convention seems to be the opposite of that approach...

If I were to reinstate mbl-partitions as config rather than a class, I think, for consistency, I'd prefer to just make both "mbl-partitions" and "mbl-artifact-names" config files.

Reply:
> If I were to reinstate mbl-partitions as config rather than a class, I think, for consistency, I'd prefer to just make both "mbl-partitions" and "mbl-artifact-names" config files.

I agree with this.

Reply:
Okay, I will reluctantly turn both classes into config files. @simonqhughes, are you okay with me turning the mbl-artifact-names class into a config file?

Reply:
Although... Can I put arbitrary python code into a config file that will run at the end of parse-time?

The current mbl-partitions.bbclass uses a `python() __anonymous {` block to do some validation, but in the near future I'll need to do something more sophisticated.

Reply:
oe-core:meta/conf/machine/include/arm/feature-arm-thumb.inc seems to be using an anonymous python block.

Reply:
But then again, if I turned mbl-artifact-names into a config file, what would `include`/`require` it? It's not really MACHINE config... I'm going to use `INHERIT += "mbl-artifact-names"` in mbl-partitions.conf.

Reply:
Nope - BitBake doesn't like
```
python __anonymous() {
```
or
```
python () {
```
in conf files. Don't know what's happening in oe-core:meta/conf/machine/include/arm/feature-arm-thumb.inc.

I'm now back to proposing that I leave this PR as-is.

Reply:
Have you tried using `.inc` instead of `.conf`?

Maybe you can use the following as inspiration:
http://arago-project.org/git/?p=meta-arago.git;a=blob;f=meta-arago-extras/conf/distro/include/tcmode-external-arm.inc;h=4074d6dd6d5a8d4ac17fe834c66f72b57088bcb9;hb=HEAD

Reply:
According to the manual, you can only define functions in class (.bbclass) and recipe (.bb or .inc) files :
https://www.yoctoproject.org/docs/1.6/bitbake-user-manual/bitbake-user-manual.html#functions

Reply:
Okay, so the real issue is whether I can get a function to run automatically at the end of parse time from a conf/inc file, not whether I can define a function at all... Although I suppose I could get a function called by using [inline Python variable expansion](https://www.yoctoproject.org/docs/1.6/bitbake-user-manual/bitbake-user-manual.html#inline-python-variable-expansion) :  `SOME_VAR = "${@ my_function() }"`.

I'll take another look.

Reply:
@diego-sueiro and I spoke IRL. The variables that users are likely to actually use are the `_DIR` variables, to install e.g. config files to the right place. MBL's directory hierarchy isn't the same thing as its partition layout, though, and we shouldn't be conflating the two by using the same `_DIR` variables both the directory hierarchy and the mount points. I proposed that we move the '_DIR' variables into a different config (`require`d by mbl.conf) and keep the partition related variables and code in a bbclass to be inherited only by recipes that need partitioning info. 

@diego-sueiro, here's a comparison of this PR's branch with a new branch that moves just the `_DIR` variables into mbl-directories.conf: https://github.com/ARMmbed/meta-mbl/compare/jh-wic-param-IOTMBL-2279...jh-split-partitions-conf-IOTMBL-2279?expand=1

Reply:
I think this is the right approach.
As soon as we have a green Jenkins and Lava I'm happy to approve.

Reply:
There's a flaw in my plan :-(.

The ATF recipes will need to know what the BL3 offset is, and I can't have BSP recipes inheriting from a class in meta-mbl-distro... And this info really is distro specific.

I think the only way to deal with this is to put the info that the ATF recipes need in the distro config.

I suppose the right way to think about it is actually that the ATF recipes "own" a variable for the BL3 offset (and give it a default value) and that the distro config overrides the value of that variable.

I think separating the directory and partition related variables was good, but I'll need to `INHERIT += "mbl-partitions"` after all.

As I discussed with Diego earlier, this has an annoying side effect: every recipe inherits the partition variable validation code, so when there's a validation error, every recipe complains... And there may be several recipes being parsed concurrently so even if the error is `fatal` it can get reported many times.

It doesn't make sense to separate the code for setting the variable values and the validation code so I suppose we'll just have to live with that.

Reply:
https://github.com/ARMmbed/meta-mbl/pull/610/commits/d25c62cbe4baebd228a4b11b40e5d95d1a79a911